File: WOT-iOS/WOT-iOS/core/datasources/WOTTree/WOTTree.m
//
//  WOTTree.m
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 6/29/15.
//  Copyright (c) 2015 Pavel Yeshchyk. All rights reserved.
//

#import "WOTTree.h"

@interface WOTTree ()

@property (nonatomic, strong)NSMutableSet *rootNodes_;
@property (nonatomic, strong)NSMutableDictionary *levelIndex;

@end

@implementation WOTTree

- (void)dealloc {
    
    [self.rootNodes_ removeAllObjects];
}

- (NSSet *)rootNodes {
    
    return self.rootNodes_;
}

- (void)addNode:(WOTNode *)node {
    
    if (!self.rootNodes_) {
        
        self.rootNodes_ = [[NSMutableSet alloc] init];
    }
    
    [self.rootNodes_ addObject:node];
    [self reindex];
}

- (void)removeAllNodes {
    
    [self.rootNodes_ removeAllObjects];
    [self reindex];
    
}
- (void)removeNode:(WOTNode *)node {
    
    [self.rootNodes_ removeObject:node];
    [self reindex];
}

- (NSArray *)nodes {
    
    if (self.nodeComparator) {
        
        return [[self.rootNodes_ allObjects] sortedArrayUsingComparator:self.nodeComparator];
    } else {
        
        return [self.rootNodes_ allObjects];
    }
}

- (void)reindex {
    
    [self.levelIndex removeAllObjects];
    self.levelIndex = [[NSMutableDictionary alloc] init];
    NSInteger level = 0;

    NSArray *rootNodes = [self nodes:[self.rootNodes allObjects] sortedUsingComparator:self.nodeComparator];
    [rootNodes enumerateObjectsUsingBlock:^(WOTNode *node, NSUInteger idx, BOOL *stop) {
        
        [self reindexChildNode:node atLevel:level];
    }];
}

- (WOTNode *)nodeAtIndexPath:(NSIndexPath *)indexPath {
    
    NSArray *section = self.levelIndex[@(indexPath.section)];
    return section[indexPath.row];
}

- (NSInteger)nodesCountAtSection:(NSInteger)sectionIndex {
    
    NSArray *section = self.levelIndex[@(sectionIndex)];
    return [section count];
}

- (void)reindexChildNode:(WOTNode *)node atLevel:(NSInteger)level{
    
    NSMutableArray *items = [NSMutableArray arrayWithArray:self.levelIndex[@(level)]];
    [items addObject:node];
    self.levelIndex[@(level)] = [items copy];

    NSArray *childNodes = [self nodes:node.children sortedUsingComparator:self.nodeComparator];
    [childNodes enumerateObjectsUsingBlock:^(WOTNode *childNode, NSUInteger idx, BOOL *stop) {
        
        [self reindexChildNode:childNode atLevel:level+1];
    }];
    
}


- (NSArray *)nodes:(NSArray *)nodes sortedUsingComparator:(WOTNodeComparator)comparator {

    if (comparator) {
        return [nodes sortedArrayUsingComparator:comparator];
    } else {
        return nodes;
    }
}

- (NSInteger)levels {
    
    return [[self.levelIndex allKeys] count];
}

- (NSInteger)width {
    
    __block NSInteger result = 0;
    [[self.levelIndex allKeys] enumerateObjectsUsingBlock:^(id key, NSUInteger idx, BOOL *stop) {
        
        id arr = self.levelIndex[key];
        result = MAX(result,[arr count]);
    }];
    return result;
}

- (NSInteger)endpointsCount {

    __block NSInteger result = 0;
    [self.rootNodes_ enumerateObjectsUsingBlock:^(WOTNode *node, BOOL *stop) {
    
        result += [self endPointsCountForNode:node];
    }];
    return result;
}

- (NSInteger)endPointsCountForNode:(WOTNode *)node {

    NSInteger cnt = [node.children count];

    if (cnt == 0) {
        
        return 1;
    }
    
    NSInteger result = 0;
    for (int i=0;i<cnt;i++) {
        
        WOTNode *childNode = node.children[i];
        result += [self endPointsCountForNode:childNode];
    }
    
    return result;
    
}

- (NSInteger)childrenCountForSiblingNode:(WOTNode *)node {
    
    __block NSInteger result = 0;
    WOTNode *parent = node.parent;
    if (parent) {
        
        NSInteger indexOfNode = [parent.children indexOfObject:node];
        for (int i=0;i<indexOfNode;i++) {
            
            WOTNode *child = parent.children[i];
            result += [self endPointsCountForNode:child];
        }

        result += [self childrenCountForSiblingNode:parent];
    }
    
    return result;
}

@end

File: WOT-iOS/WOT-iOS/core/datasources/WOTTree/WOTTree.swift
//
//  WOTTree.swift
//  WOT-iOS
//
//  Created by Pavel Yeshchyk on 7/16/18.
//  Copyright © 2018 Pavel Yeshchyk. All rights reserved.
//

import Foundation

//typealias WOTNodeComparator = (_ left: WOTNodeProtocol, _ right: WOTNodeProtocol) -> ComparisonResult

@objc
protocol WOTTreeProtocol: NSObjectProtocol {
    var model: WOTDataModel { get set }
    var nodes: [WOTNodeProtocol] { get }
    var levels: Int { get }
    var width: Int { get }
    var endpointsCount: Int { get }
//    var nodeComparator: WOTNodeComparator { get set }
    var rootNodes: [WOTNodeProtocol] { get }
    func nodesCount(sectionIndex: Int) -> Int
    func node(atIndexPath: NSIndexPath) -> WOTNodeProtocol?
    func addNode(node: WOTNodeProtocol)
    func removeNode(node: WOTNodeProtocol)
    func removeAllNodes()
    func findOrCreateRootNode(forPredicate: NSPredicate) -> WOTNodeProtocol

}

@objc
public class WOTTreeSwift: NSObject {

    var model: WOTDataModel

    @objc
    var levels: Int {
        return self.model.levels
    }

    @objc
    var width: Int {
        return self.model.width
    }

    @objc
    var endpointsCount: Int {
        return self.model.endpointsCount
    }

    @objc
    var rootNodes: [WOTNodeProtocol] {
        return self.model.rootNodes
    }

    @objc
    var nodes: [WOTNodeProtocol] {
        return self.model.allObjects(sortComparator: { (_ , _) -> Bool in
            return true
        })
    }
    
    @objc
    init(dataModel: WOTDataModel) {
        model = dataModel
    }

    @objc
    func nodesCount(sectionIndex: Int) -> Int {
        return self.model.nodesCount(section:sectionIndex)
    }

    @objc
    func node(atIndexPath: NSIndexPath) -> WOTNodeProtocol? {
        return self.model.node(atIndexPath: atIndexPath)
    }

    func addNode(node: WOTNodeProtocol) {
        self.model.add(node: node)
    }

    func removeNode(node: WOTNodeProtocol) {
        self.model.remove(node: node)
    }

    func removeAllNodes() {
        self.model.removeAll()
    }

    func findOrCreateRootNode(forPredicate: NSPredicate) -> WOTNodeProtocol {
        let roots = self.rootNodes.filter {_ in forPredicate.evaluate(with: nil)}
        if roots.count == 0 {
            let root = WOTNodeSwift(name: "root")
            self.addNode(node: root)
            return root
        } else {
            let root = roots.first
            return root!
        }
    }

}
